tsl/src/continuous_aggs/invalidation_plugin.c

```diff
@@ -0,0 +1,552 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <fmgr.h>
+
+#include <inttypes.h>
+#include <miscadmin.h>
+
+#include <access/attnum.h>
+#include <access/genam.h>
+#include <access/htup.h>
+#include <access/table.h>
+#include <access/transam.h>
+#include <access/tupdesc.h>
+#include <catalog/namespace.h>
+#include <catalog/pg_attribute.h>
+#include <catalog/pg_inherits.h>
+#include <commands/defrem.h>
+#include <executor/tuptable.h>
+#include <nodes/makefuncs.h>
+#include <replication/logical.h>
+#include <replication/logicalproto.h>
+#include <replication/output_plugin.h>
+#include <storage/lockdefs.h>
+#include <utils/builtins.h>
+#include <utils/fmgroids.h>
+#include <utils/hsearch.h>
+#include <utils/lsyscache.h>
+#include <utils/memutils.h>
+#include <utils/palloc.h>
+#include <utils/regproc.h>
+#include <utils/varlena.h>
+
+#include "export.h"
+#include "invalidation_plugin_cache.h"
+#include "ts_catalog/catalog.h"
+
+PG_MODULE_MAGIC;
+
+#if PG16_LT
+extern TSDLLEXPORT void _PG_output_plugin_init(OutputPluginCallbacks *cb);
+#endif
+
+typedef struct TrackRelInfo
+{
+	RangeVar *hypertable_rv;
+	Oid hypertable_relid; /* Relid of the hypertable that we're collecting
+						   * invalidations for */
+	NameData attname;	  /* Name of the attribute to read from the chunks */
+	int attnum;			  /* Attribute number of the attribute to read from
+						   * the chunks. */
+} TrackRelInfo;
+
+/*
+ * Hash table with relations seen.
+ *
+ * In order to avoid multiple lookups of a relation, we record all relations
+ * seen as entries in a hash table.
+ *
+ * If we see a new relation, we look up the inheritance parent and if it has
+ * one, we check if it is one of the relations that we should track and set
+ * the "track" field to true in that case.
+ *
+ * If the relation does not have an inheritance parent, or if the inheritance
+ * parent is not one of the relations we track, we will set the "track" field
+ * to false.
+ *
+ * Note that since hypertables are single-level inheritance tables, we do not
+ * look further up the inheritance chain. If you want to be able to handle
+ * multi-level inheritance changes, you need to update this.
+ */
+typedef struct SeenRelsEntry
+{
+	Oid relid;			   /* Relid, also key */
+	bool track;			   /* Shall this relation be tracked or not? */
+	TrackRelInfo *relinfo; /* Relation information, or NULL if not tracked */
+} SeenRelsEntry;
+
+typedef struct InvalidationsPluginData
+{
+	MemoryContext context;
+	Relation logrel;	/* The log relation used for tuples passed back to
```

Comment:
Looking further down the code, this rel is opened and reference is held in the plugin but it looks like the rel is only used in order to call a logical replication API that serializes a HeapTuple in format of that rel. It doesn't seem like we actually need this rel (see other comments on serialization), because the `logicalrep_write_insert` only uses the rel to read the TupleDesc. We can skip all this "rel" stuff and just serialize directly with pq_send<type> functions.

Reply:
This is gone now, so no longer relevant.